components/lcd/esp_lcd_st7796/esp_lcd_st7796.c

```diff
@@ -1,353 +1,36 @@
-
-    if (st7796->reset_gpio_num >= 0) {
-        gpio_reset_pin(st7796->reset_gpio_num);
-    }
-    ESP_LOGD(TAG, "del st7796 panel @%p", st7796);
-    free(st7796);
-    return ESP_OK;
-}
-
-static esp_err_t panel_st7796_reset(esp_lcd_panel_t *panel)
-{
-    st7796_panel_t *st7796 = __containerof(panel, st7796_panel_t, base);
-    esp_lcd_panel_io_handle_t io = st7796->io;
-
-    // perform hardware reset
-    if (st7796->reset_gpio_num >= 0) {
-        gpio_set_level(st7796->reset_gpio_num, st7796->reset_level);
-        vTaskDelay(pdMS_TO_TICKS(10));
-        gpio_set_level(st7796->reset_gpio_num, !st7796->reset_level);
-        vTaskDelay(pdMS_TO_TICKS(120));
-    } else { // perform software reset
-        ESP_RETURN_ON_ERROR(esp_lcd_panel_io_tx_param(io, LCD_CMD_SWRESET, NULL, 0), TAG, "send command failed");
-        vTaskDelay(pdMS_TO_TICKS(120)); // spec, wait at least 5ms before sending new command
-    }
-
-    return ESP_OK;
-}
-
-static const st7796_lcd_init_cmd_t vendor_specific_init_default[] = {
-//  {cmd, { data }, data_size, delay_ms}
-    {0xf0, (uint8_t []){0xc3}, 1, 0},
-    {0xf0, (uint8_t []){0x96}, 1, 0},
-    {0xb4, (uint8_t []){0x01}, 1, 0},
-    {0xb7, (uint8_t []){0xc6}, 1, 0},
-    {0xe8, (uint8_t []){0x40, 0x8a, 0x00, 0x00, 0x29, 0x19, 0xa5, 0x33}, 8, 0},
-    {0xc1, (uint8_t []){0x06}, 1, 0},
-    {0xc2, (uint8_t []){0xa7}, 1, 0},
-    {0xc5, (uint8_t []){0x18}, 1, 0},
-    {0xe0, (uint8_t []){0xf0, 0x09, 0x0b, 0x06, 0x04, 0x15, 0x2f, 0x54, 0x42, 0x3c, 0x17, 0x14, 0x18, 0x1b}, 14, 0},
-    {0xe1, (uint8_t []){0xf0, 0x09, 0x0b, 0x06, 0x04, 0x03, 0x2d, 0x43, 0x42, 0x3b, 0x16, 0x14, 0x17, 0x1b}, 14, 0},
-    {0xf0, (uint8_t []){0x3c}, 1, 0},
-    {0xf0, (uint8_t []){0x69}, 1, 0},
-};
-
-static esp_err_t panel_st7796_init(esp_lcd_panel_t *panel)
-{
-    st7796_panel_t *st7796 = __containerof(panel, st7796_panel_t, base);
-    esp_lcd_panel_io_handle_t io = st7796->io;
-
-    // LCD goes into sleep mode and display will be turned off after power on reset, exit sleep mode first
-    ESP_RETURN_ON_ERROR(esp_lcd_panel_io_tx_param(io, LCD_CMD_SLPOUT, NULL, 0), TAG, "send command failed");
-    vTaskDelay(pdMS_TO_TICKS(100));
-    ESP_RETURN_ON_ERROR(esp_lcd_panel_io_tx_param(io, LCD_CMD_MADCTL, (uint8_t[]) {
-        st7796->madctl_val,
-    }, 1), TAG, "send command failed");
-    ESP_RETURN_ON_ERROR(esp_lcd_panel_io_tx_param(io, LCD_CMD_COLMOD, (uint8_t[]) {
-        st7796->colmod_val,
-    }, 1), TAG, "send command failed");
-
-    const st7796_lcd_init_cmd_t *init_cmds = NULL;
-    uint16_t init_cmds_size = 0;
-    if (st7796->init_cmds) {
-        init_cmds = st7796->init_cmds;
-        init_cmds_size = st7796->init_cmds_size;
     } else {
-        init_cmds = vendor_specific_init_default;
-        init_cmds_size = sizeof(vendor_specific_init_default) / sizeof(st7796_lcd_init_cmd_t);
+        ret = esp_lcd_new_panel_st7796_general(io, panel_dev_config, ret_panel);
```

Comment:
this is the so-called "runtime dispatch", where we will always link in the SPI and I80 driver, even user only uses the MIPI interface. Leading to a significant increment to the binary size.

Reply:
but it's up to you, not a blocker of this PR

Reply:
Yes, I understand this situation. This will indeed add extra binary size, but it is done for compatibility with previous versions. I don't want this change to become a breaking change. I will consider separating them in the future.